--- gentoo-sources.ebuild	2024-05-13 18:10:15.367002597 -0000
+++ gentoo-sources-ps3.ebuild	2024-05-13 21:04:27.834719495 -0000
@@ -10,6 +10,7 @@
 detect_version
 detect_arch
 
-DESCRIPTION="Full sources including the Gentoo patchset for the ${KV_MAJOR}.${KV_MINOR} kernel tree"
-HOMEPAGE="https://dev.gentoo.org/~mpagano/genpatches"
+DESCRIPTION="Full sources including the Gentoo patchset for the ${KV_MAJOR}.${KV_MINOR} kernel tree and PS3 patches"
+HOMEPAGE="https://github.com/damiandudycz/ps3"
+PS3_URI="https://raw.githubusercontent.com/damiandudycz/ps3-gentoo-overlay.distfiles/main/sys-kernel/gentoo-sources-ps3/gentoo-sources-ps3-files-${PVR}.tar.xz"
 SRC_URI="${KERNEL_URI} ${GENPATCHES_URI} ${ARCH_URI}"
@@ -15,1 +16,1 @@
-SRC_URI="${KERNEL_URI} ${GENPATCHES_URI} ${ARCH_URI}"
+SRC_URI="${KERNEL_URI} ${GENPATCHES_URI} ${ARCH_URI} ${PS3_URI}"
@@ -25,3 +26,17 @@
 pkg_postrm() {
 	kernel-2_pkg_postrm
 }
+
+src_prepare() {
+einfo "WORKDIR: ${WORKDIR}"
+einfo "DISTDIR: ${DISTDIR}"
+    default
+    unpack "gentoo-sources-ps3-files-${PVR}.tar.xz"
+    mv "${S}/ps3_gentoo_defconfig" "${S}/arch/powerpc/configs/ps3_defconfig" || die
+    for patch in "${S}/ps3_patches"/*.patch; do
+        eapply "${patch}"
+    done
+    # Cleanup helper files
+    rm -f "${S}/ps3_defconfig_diffs"
+    rm -rf "${S}/ps3_patches"
+}
