Fixes lv1 calls in PS3 repository module.

diff --git a/arch/powerpc/include/asm/lv1call.h b/arch/powerpc/include/asm/lv1call.h
index 5cddd73..e6563c6 100644
--- a/arch/powerpc/include/asm/lv1call.h
+++ b/arch/powerpc/include/asm/lv1call.h
@@ -254,10 +254,10 @@ LV1_CALL(get_logical_partition_id,                      0, 1,  74 )
 LV1_CALL(configure_execution_time_variable,             1, 0,  77 )
 LV1_CALL(get_spe_irq_outlet,                            2, 1,  78 )
 LV1_CALL(set_spe_privilege_state_area_1_register,       3, 0,  79 )
-LV1_CALL(create_repository_node,                        6, 0,  90 )
+LV1_CALL(create_repository_node,                        7, 0,  90 )
 LV1_CALL(read_repository_node,                          5, 2,  91 )
-LV1_CALL(write_repository_node,                         6, 0,  92 )
-LV1_CALL(delete_repository_node,                        4, 0,  93 )
+LV1_CALL(write_repository_node,                         7, 0,  92 )
+LV1_CALL(delete_repository_node,                        5, 0,  93 )
 LV1_CALL(read_htab_entries,                             2, 5,  95 )
 LV1_CALL(set_dabr,                                      2, 0,  96 )
 LV1_CALL(undocumented_function_99,                      2, 0,  99 )
diff --git a/arch/powerpc/platforms/ps3/repository.c b/arch/powerpc/platforms/ps3/repository.c
index 1abe33f..28ef4da 100644
--- a/arch/powerpc/platforms/ps3/repository.c
+++ b/arch/powerpc/platforms/ps3/repository.c
@@ -1052,7 +1052,7 @@ static int create_node(u64 n1, u64 n2, u64 n3, u64 n4, u64 v1, u64 v2)
 
 	dump_node(0, n1, n2, n3, n4, v1, v2);
 
-	result = lv1_create_repository_node(n1, n2, n3, n4, v1, v2);
+	result = lv1_create_repository_node(n1, n2, n3, n4, v1, v2, 0);
 
 	if (result) {
 		pr_devel("%s:%d: lv1_create_repository_node failed: %s\n",
@@ -1069,7 +1069,7 @@ static int delete_node(u64 n1, u64 n2, u64 n3, u64 n4)
 
 	dump_node(0, n1, n2, n3, n4, 0, 0);
 
-	result = lv1_delete_repository_node(n1, n2, n3, n4);
+	result = lv1_delete_repository_node(n1, n2, n3, n4, 0);
 
 	if (result) {
 		pr_devel("%s:%d: lv1_delete_repository_node failed: %s\n",
@@ -1089,7 +1089,7 @@ static int write_node(u64 n1, u64 n2, u64 n3, u64 n4, u64 v1, u64 v2)
 	if (!result)
 		return 0;
 
-	result = lv1_write_repository_node(n1, n2, n3, n4, v1, v2);
+	result = lv1_write_repository_node(n1, n2, n3, n4, v1, v2, 0);
 
 	if (result) {
 		pr_devel("%s:%d: lv1_write_repository_node failed: %s\n",
