diff --git a/arch/powerpc/include/asm/ps3.h b/arch/powerpc/include/asm/ps3.h
index 83a633a..a2625ab 100644
--- a/arch/powerpc/include/asm/ps3.h
+++ b/arch/powerpc/include/asm/ps3.h
@@ -340,6 +340,7 @@ enum ps3_match_sub_id {
 #define PS3_MODULE_ALIAS_STOR_NOR_FLASH	"ps3:12:0"
 #define PS3_MODULE_ALIAS_DISP_MANAGER	"ps3:13:0"
 #define PS3_MODULE_ALIAS_STOR_ENCDEC	"ps3:14:0"
+#define PS3_MODULE_ALIAS_JUPITER	"ps3:15:0"
 
 enum ps3_system_bus_device_type {
 	PS3_DEVICE_TYPE_IOC0 = 1,
diff --git a/drivers/net/wireless/ps3jupiter/ps3_jupiter.c b/drivers/net/wireless/ps3jupiter/ps3_jupiter.c
index 7f65b6c..e26bdcd 100644
--- a/drivers/net/wireless/ps3jupiter/ps3_jupiter.c
+++ b/drivers/net/wireless/ps3jupiter/ps3_jupiter.c
@@ -1238,8 +1238,8 @@ static void __exit ps3_jupiter_exit(void)
 module_init(ps3_jupiter_init);
 module_exit(ps3_jupiter_exit);
 
-MODULE_SUPPORTED_DEVICE("PS3 Jupiter");
 MODULE_DEVICE_TABLE(usb, ps3_jupiter_devtab);
 MODULE_DESCRIPTION("PS3 Jupiter");
 MODULE_AUTHOR("glevand");
 MODULE_LICENSE("GPL");
+MODULE_ALIAS(PS3_MODULE_ALIAS_JUPITER);
diff --git a/drivers/net/wireless/ps3jupiter/ps3_jupiter_sta.c b/drivers/net/wireless/ps3jupiter/ps3_jupiter_sta.c
index b19de88..2346ed3 100644
--- a/drivers/net/wireless/ps3jupiter/ps3_jupiter_sta.c
+++ b/drivers/net/wireless/ps3jupiter/ps3_jupiter_sta.c
@@ -2927,8 +2927,8 @@ static void __exit ps3_jupiter_sta_exit(void)
 module_init(ps3_jupiter_sta_init);
 module_exit(ps3_jupiter_sta_exit);
 
-MODULE_SUPPORTED_DEVICE("PS3 Jupiter STA");
 MODULE_DEVICE_TABLE(usb, ps3_jupiter_sta_devtab);
 MODULE_DESCRIPTION("PS3 Jupiter STA");
 MODULE_AUTHOR("glevand");
 MODULE_LICENSE("GPL");
+MODULE_ALIAS("ps3:15:0");
