Adds lv1 calls:
 - 105, 106, 107, 108, 109 - for UDBG
 - 114, 115 - for PS3PHYSMEM

diff --git a/arch/powerpc/include/asm/lv1call.h b/arch/powerpc/include/asm/lv1call.h
index b2f19a698..85f141efb 100644
--- a/arch/powerpc/include/asm/lv1call.h
+++ b/arch/powerpc/include/asm/lv1call.h
@@ -261,6 +261,13 @@ LV1_CALL(delete_repository_node,                        4, 0,  93 )
 LV1_CALL(read_htab_entries,                             2, 5,  95 )
 LV1_CALL(set_dabr,                                      2, 0,  96 )
 LV1_CALL(get_total_execution_time,                      2, 1, 103 )
+LV1_CALL(undocumented_function_105,                     7, 0, 105 )
+LV1_CALL(undocumented_function_106,                     1, 0, 106 )
+LV1_CALL(undocumented_function_107,                     6, 1, 107 )
+LV1_CALL(undocumented_function_108,                     1, 5, 108 )
+LV1_CALL(undocumented_function_109,                     1, 0, 109 )
+LV1_CALL(undocumented_function_114,                     3, 1, 114 )
+LV1_CALL(undocumented_function_115,                     1, 0, 115 )
 LV1_CALL(allocate_io_segment,                           3, 1, 116 )
 LV1_CALL(release_io_segment,                            2, 0, 117 )
 LV1_CALL(construct_io_irq_outlet,                       1, 1, 120 )
