Adds lv1 hypervisor calls 250, 251, 252, 253 for region management.

diff --git a/arch/powerpc/include/asm/lv1call.h b/arch/powerpc/include/asm/lv1call.h
index 85f141efb..de789361b 100644
--- a/arch/powerpc/include/asm/lv1call.h
+++ b/arch/powerpc/include/asm/lv1call.h
@@ -323,6 +323,10 @@ LV1_CALL(storage_read,                                  6, 1, 245 )
 LV1_CALL(storage_write,                                 6, 1, 246 )
 LV1_CALL(storage_send_device_command,                   6, 1, 248 )
 LV1_CALL(storage_get_async_status,                      1, 2, 249 )
+LV1_CALL(storage_create_region,                         5, 2, 250 )
+LV1_CALL(storage_delete_region,                         2, 1, 251 )
+LV1_CALL(storage_set_region_acl,                        4, 1, 252 )
+LV1_CALL(storage_get_region_acl,                        3, 2, 253 )
 LV1_CALL(storage_check_async_status,                    2, 1, 254 )
 LV1_CALL(panic,                                         1, 0, 255 )
 LV1_CALL(construct_lpm,                                 6, 3, 140 )
